File: packages/lynx-deep-linking/ios/src/MiniLynxDeepLinkingModule.m
#import "MiniLynxDeepLinkingModule.h"

#import <Lynx/LynxContext.h>

#import "MiniLynxDeepLinkingCenter.h"

@interface MiniLynxDeepLinkingModule ()
@property(nonatomic, weak) LynxContext *context;
@end

@implementation MiniLynxDeepLinkingModule

+ (NSString *)name {
  return @"MiniLynxDeepLinkingModule";
}

/**
 * The JavaScript-name-to-selector table Lynx dispatches through.
 *
 * It is hand-written and therefore hand-checked: a name here that no selector
 * below implements fails at call time rather than at build time, and a selector
 * whose arity does not match what JavaScript passes fails the same way. Adding
 * a method means editing this table and the TypeScript facade together.
 */
+ (NSDictionary<NSString *, NSString *> *)methodLookup {
  return @{
    @"getInitialURL" : NSStringFromSelector(@selector(getInitialURL:)),
    @"flushPendingLink" : NSStringFromSelector(@selector(flushPendingLink)),
    @"openURL" : NSStringFromSelector(@selector(openURL:callback:)),
    @"canOpenURL" : NSStringFromSelector(@selector(canOpenURL:callback:)),
    @"openSettings" : NSStringFromSelector(@selector(openSettings:)),
  };
}

- (instancetype)initWithLynxContext:(LynxContext *)context {
  if (self = [super init]) {
    _context = context;
    [MiniLynxDeepLinkingCenter.shared registerContext:context];
  }
  return self;
}

- (void)dealloc {
  if (_context) [MiniLynxDeepLinkingCenter.shared unregisterContext:_context];
}

#pragma mark - Inbound

- (void)getInitialURL:(void (^)(id))callback {
  NSString *url = MiniLynxDeepLinkingCenter.shared.initialURL;
  // No launch URL is a null rather than an omission: `getInitialURL` is typed
  // `string | null` and an absent argument would arrive as `undefined`.
  callback(url ?: NSNull.null);
}

- (void)flushPendingLink {
  [MiniLynxDeepLinkingCenter.shared flushPendingLink];
}

#pragma mark - Outbound

- (void)openURL:(NSString *)url callback:(void (^)(id))callback {
  [MiniLynxDeepLinkingCenter.shared openURL:url
                                 completion:^(NSDictionary *result) {
                                   callback(result);
                                 }];
}

- (void)canOpenURL:(NSString *)url callback:(void (^)(id))callback {
  [MiniLynxDeepLinkingCenter.shared canOpenURL:url
                                    completion:^(BOOL canOpen) {
                                      callback(@(canOpen));
                                    }];
}

- (void)openSettings:(void (^)(id))callback {
  [MiniLynxDeepLinkingCenter.shared openSettingsWithCompletion:^(BOOL opened) {
    callback(@(opened));
  }];
}

@end

File: packages/lynx-deep-linking/android/src/main/java/dev/amritk/minilynx/deeplinking/MiniLynxDeepLinkingModule.kt
package dev.amritk.minilynx.deeplinking

import com.lynx.jsbridge.LynxContextModule
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxNativeModule
import com.lynx.react.bridge.Callback
import com.lynx.tasm.behavior.LynxContext

/**
 * The Android half of `@amritk/lynx-deep-linking`.
 *
 * Every method here is the other end of one function in the package's
 * TypeScript. The names, arities and call forms have to match exactly:
 * `src/testing/create-fake-deep-linking.ts` is the executable statement of that
 * contract, and it is what the JavaScript suite runs against.
 *
 * `@LynxNativeModule` registers the module through Lynx's annotation processor,
 * so an app that autolinks this library gets it without writing anything. An
 * app whose build does not run the processor can register it by hand instead:
 *
 * ```kotlin
 * LynxEnv.inst().registerModule("MiniLynxDeepLinkingModule", MiniLynxDeepLinkingModule::class.java)
 * ```
 *
 * ## The inbound direction is not here
 *
 * Nothing in this class receives a link. Intents arrive at an activity, long
 * before any LynxView exists and on a lifecycle this module cannot see, so
 * `DeepLinks` owns the entry points and `DeepLinkEvents` owns the fan-out. This
 * is only the JavaScript-facing surface: three questions and one flush.
 */
@LynxNativeModule(name = "MiniLynxDeepLinkingModule")
class MiniLynxDeepLinkingModule(context: LynxContext) : LynxContextModule(context) {

  init {
    // The context is what `sendGlobalEvent` needs, and a module instance is
    // per-LynxView, so registering here is what lets links reach every view
    // that is up rather than whichever one happened to be built last.
    DeepLinkEvents.register(context)
  }

  @LynxMethod
  fun getInitialURL(callback: Callback) {
    val url = DeepLinks.initialURL()
    if (url == null) callback.invoke(*NO_URL) else callback.invoke(url)
  }

  /**
   * Replays a link that arrived with no view to receive it.
   *
   * Returning rather than callback-shaped: there is nothing to wait for, and
   * the facade calls it through `callNative` immediately after subscribing —
   * being asked is the only reliable signal that JavaScript is listening. See
   * `DeepLinkEvents`.
   */
  @LynxMethod
  fun flushPendingLink() {
    DeepLinkEvents.flushPendingLink()
  }

  @LynxMethod
  fun openURL(url: String, callback: Callback) {
    callback.invoke(SystemLinks.open(mLynxContext.applicationContext, url))
  }

  @LynxMethod
  fun canOpenURL(url: String, callback: Callback) {
    callback.invoke(SystemLinks.canOpen(mLynxContext.applicationContext, url))
  }

  @LynxMethod
  fun openSettings(callback: Callback) {
    callback.invoke(SystemLinks.openSettings(mLynxContext.applicationContext))
  }

  companion object {
    /**
     * `Callback.invoke` takes `Object...`, and a bare Kotlin `null` there is
     * read as a null *array* rather than as one null argument. A pre-built
     * single-null array is the unambiguous spelling — and `getInitialURL` is
     * typed `string | null`, so the difference between a null and an absent
     * argument is the difference between `null` and `undefined` on the other
     * side.
     */
    private val NO_URL = arrayOfNulls<Any>(1)
  }
}
